### PR TITLE
feat(module/vpn): VPN module for GCP VM-Series deployments

### PR DESCRIPTION
**Pull request by pavelrn**
## Description

Introducing VPN module for GCP VM-Series deployments. The module allows to deploy: 
- VPN HA gateway with Cloud Router,
- multiple VPN instances (connections) where each instance is represented by remote gateway and multiple VPN tunnels.

PR closes [Issue 215](https://github.com/PaloAltoNetworks/terraform-google-vmseries-modules/issues/215).

## Motivation and Context

VPN module is required for building GCP to on-prem connectivity, for GCP to GCP connectivity.

## How Has This Been Tested?

Terraform plan/apply using example VPN configuration.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes if appropriate.
- [x] All new and existing tests passed.

### CODE REVIEW

File: modules/vpn/example/main.tf [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  source = "../"
+
+  project = "gcp-gcs-pso"
+  region  = "us-central1"
+
+  vpn_gateway_name = "my-test-gateway"
+  router_name      = "my-test-router"
+  network          = data.google_compute_network.test.self_link
+
+  vpn_config = {
```
Comment (horiagunica) [inline modules/vpn/example/main.tf:16]: Maybe we can move some of this config into tfvars ?

File: modules/vpn/modules/vpn_instance/main.tf [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  peer_external_gateway = (
+    var.peer_external_gateway != null
+    ? google_compute_external_vpn_gateway.external_gateway[0].self_link
+    : null
+  )
+  secret = random_id.secret.b64_url
+}
+
+resource "google_compute_router" "router" {
+  name    = coalesce(var.router_name, "rtr-${var.name}")
```
Comment (horiagunica) [inline modules/vpn/modules/vpn_instance/main.tf:11]: Maybe we can modify the `var.name` variable into `name_prefix` to be aligned with the other modules/examples ?
Also - the position of this variable here and in subsequent places should be at the beginning of the resource name.

File: modules/vpn/modules/vpn_instance/variables.tf [hunk 3]
```diff
@@ -0,0 +1,123 @@
+variable "name" {
```
Comment (horiagunica) [inline modules/vpn/modules/vpn_instance/variables.tf:1]: Modify the variable name to `name_prefix`

**Comment by horiagunica** [comment]
I think the folder structure should be aligned with our current structure :
- Move example to `examples` folder
- I can see there is an inherited module - maybe it would be a good idea to flatten that module altogether and have them all in a single `main.tf` file.

File: modules/vpn/example/example.tfvars [hunk 4]
```diff
@@ -0,0 +1,129 @@
+project      = "<project_id>"
+region       = "us-central1"
+network_name = "<network_name>"
+
+vpn_gateway_name = "my-test-gateway"
+router_name      = "my-test-router"
+
+vpn_config = {
```
Comment (horiagunica) [inline modules/vpn/example/example.tfvars:8]: This config is also in `main.tf` - it needs to be aligned somehow - it should be taken from main.tf and moved here solely .

**Review by horiagunica — DISMISSED** [review]

File: modules/vpn/example/main.tf (hunk 1 above)
Comment (pavelrn) [reply modules/vpn/example/main.tf]: Sure. That's an obvious oversight on my side. Fixed

File: modules/vpn/example/example.tfvars (hunk 4 above)
Comment (pavelrn) [reply modules/vpn/example/example.tfvars:8]: Sure. That's an obvious oversight on my side. Fixed

**Comment by pavelrn** [comment]
@horiagunica 

> * Move example to `examples` folder

The reason why I didn't place the example into main `examples` folder is that the module itself is NOT a VM-Series example. I still believe that VPN module example doesn't belong to main `examples` folder.

> * I can see there is an inherited module - maybe it would be a good idea to flatten that module altogether and have them all in a single `main.tf` file.

The reason why I placed `vpn_instance` in `vpn` module is that because 
-`vpn_instance` is called from `vpn` and 
-`vpn_instance` is not supposed to be used (will not be used) without `vpn`. 
I still believe we should place `vpn_instance` inside `vpn` not to expose the modules that are not supposed to be used as external.

File: modules/vpn/modules/vpn_instance/main.tf (hunk 2 above)
Comment (pavelrn) [reply modules/vpn/modules/vpn_instance/main.tf]: Fixed

**Comment by pavelrn** [comment]
I decided to flatten the module (remove submodule) and put example code in the module's README

**Review by migara — APPROVED** [review]

**Pull request merged**